File: src/main/java/com/antwerkz/sofia/Ophelia.java
package com.antwerkz.sofia;

import java.io.*;
import java.net.*;
import java.text.*;
import java.util.*;
import java.security.*;
import java.util.ResourceBundle.Control;

import org.slf4j.*;

public class Ophelia {
    private static Map<Locale, ResourceBundle> messages = new HashMap<>();
        private static final Logger logger = LoggerFactory.getLogger(Ophelia.class);

    private Ophelia() {}

    private static ResourceBundle getBundle(Locale... localeList) {
        Locale locale = localeList.length == 0 ? Locale.getDefault() : localeList[0];
        ResourceBundle labels = loadBundle(locale);
        if(labels == null) {
            labels = loadBundle(Locale.ROOT);
        }
        return labels;
    }

    private static ResourceBundle loadBundle(Locale locale) {
        ResourceBundle bundle = messages.get(locale);
        if(bundle == null) {
            bundle = ResourceBundle.getBundle("ophelia", locale );
            messages.put(locale, bundle);
        }
        return bundle;
    }

    private static String getMessageValue(String key, Locale... locale) {
        return (String) getBundle(locale).getObject(key);
    }

    public static String appTitle(Locale... locale) {
        return getMessageValue("app.title", locale);
    }

    public static String bookmarkExists(Locale... locale) {
        return getMessageValue("bookmark.exists", locale);
    }

    public static String bookmarks(Locale... locale) {
        return getMessageValue("bookmarks", locale);
    }

    public static String cancel(Locale... locale) {
        return getMessageValue("cancel", locale);
    }

    public static String close(Locale... locale) {
        return getMessageValue("close", locale);
    }

    public static String collectionStats(Locale... locale) {
        return getMessageValue("collectionStats", locale);
    }

    public static String collections(Locale... locale) {
        return getMessageValue("collections", locale);
    }

    public static String count(Locale... locale) {
        return getMessageValue("count", locale);
    }

    public static String explainSubmit(Locale... locale) {
        return getMessageValue("explain.submit", locale);
    }

    public static String findSubmit(Locale... locale) {
        return getMessageValue("find.submit", locale);
    }

    public static String indexes(Locale... locale) {
        return getMessageValue("indexes", locale);
    }

    public static String insertSubmit(Locale... locale) {
        return getMessageValue("insert.submit", locale);
    }

    public static String invalidQuery(Object arg0, Locale... locale) {
        return MessageFormat.format(getMessageValue("invalid.query", locale), arg0);
    }

    public static String load(Locale... locale) {
        return getMessageValue("load", locale);
    }

    public static String loadBookmark(Locale... locale) {
        return getMessageValue("load.bookmark", locale);
    }

    public static String message(Locale... locale) {
        return getMessageValue("message", locale);
    }

    public static String noResults(Locale... locale) {
        return getMessageValue("no.results", locale);
    }

    public static String operationBookmark(Locale... locale) {
        return getMessageValue("operation.bookmark", locale);
    }

    public static String operationExport(Locale... locale) {
        return getMessageValue("operation.export", locale);
    }

    public static String operationLimit(Locale... locale) {
        return getMessageValue("operation.limit", locale);
    }

    public static String operationMultiple(Locale... locale) {
        return getMessageValue("operation.multiple", locale);
    }

    public static String operationShowCount(Locale... locale) {
        return getMessageValue("operation.showCount", locale);
    }

    public static String operationUpsert(Locale... locale) {
        return getMessageValue("operation.upsert", locale);
    }

    public static String parameters(Locale... locale) {
        return getMessageValue("parameters", locale);
    }

    public static String removeSubmit(Locale... locale) {
        return getMessageValue("remove.submit", locale);
    }

    public static String resultCount(Locale... locale) {
        return getMessageValue("result.count", locale);
    }

    public static String save(Locale... locale) {
        return getMessageValue("save", locale);
    }

    public static String selectCollection(Locale... locale) {
        return getMessageValue("selectCollection", locale);
    }

    public static String unknownQueryMethod(Object arg0, Locale... locale) {
        return MessageFormat.format(getMessageValue("unknown.query.method", locale), arg0);
    }

    public static String updateSubmit(Locale... locale) {
        return getMessageValue("update.submit", locale);
    }


}

File: src/main/kotlin/com/antwerkz/sofia/Ophelia.kt
package com.antwerkz.sofia

import org.slf4j.LoggerFactory
import java.text.MessageFormat
import java.util.HashMap
import java.util.Locale
import java.util.ResourceBundle

public class Ophelia {
    private val messages = HashMap<Locale, ResourceBundle>()
    private val logger = LoggerFactory.getLogger(javaClass<Ophelia>())

    private fun getBundle(vararg localeList: Locale): ResourceBundle {
        val locale = if (localeList.size() == 0) Locale.getDefault() else localeList[0]
        var labels: ResourceBundle? = loadBundle(locale)
        if (labels == null) {
            labels = loadBundle(Locale.ROOT)
        }
        return labels!!
    }

    private fun loadBundle(locale: Locale): ResourceBundle {
        var bundle: ResourceBundle? = messages.get(locale)
        if (bundle == null) {
            bundle = ResourceBundle.getBundle("ophelia", locale)
            messages.put(locale, bundle)
        }
        return bundle!!
    }

    private fun getMessageValue(key: String, vararg locale: Locale): String {
        return getBundle(*locale).getObject(key) as String
    }

    public fun appTitle(): String {
        return getMessageValue("app.title")
    }

    public fun appTitle(locale: Locale): String {
        return getMessageValue("app.title", locale)
    }

    public fun bookmarkExists(vararg locale: Locale): String {
        return getMessageValue("bookmark.exists", *locale)
    }

    public fun bookmarks(vararg locale: Locale): String {
        return getMessageValue("bookmarks", *locale)
    }

    public fun cancel(vararg locale: Locale): String {
        return getMessageValue("cancel", *locale)
    }

    public fun close(vararg locale: Locale): String {
        return getMessageValue("close", *locale)
    }

    public fun collectionStats(vararg locale: Locale): String {
        return getMessageValue("collectionStats", *locale)
    }

    public fun collections(vararg locale: Locale): String {
        return getMessageValue("collections", *locale)
    }

    public fun count(vararg locale: Locale): String {
        return getMessageValue("count", *locale)
    }

    public fun explainSubmit(vararg locale: Locale): String {
        return getMessageValue("explain.submit", *locale)
    }

    public fun findSubmit(vararg locale: Locale): String {
        return getMessageValue("find.submit", *locale)
    }

    public fun indexes(vararg locale: Locale): String {
        return getMessageValue("indexes", *locale)
    }

    public fun insertSubmit(vararg locale: Locale): String {
        return getMessageValue("insert.submit", *locale)
    }

    public fun invalidQuery(arg0: Any, vararg locale: Locale): String {
        return MessageFormat.format(getMessageValue("invalid.query", *locale), arg0)
    }

    public fun load(vararg locale: Locale): String {
        return getMessageValue("load", *locale)
    }

    public fun loadBookmark(vararg locale: Locale): String {
        return getMessageValue("load.bookmark", *locale)
    }

    public fun message(vararg locale: Locale): String {
        return getMessageValue("message", *locale)
    }

    public fun noResults(vararg locale: Locale): String {
        return getMessageValue("no.results", *locale)
    }

    public fun operationBookmark(vararg locale: Locale): String {
        return getMessageValue("operation.bookmark", *locale)
    }

    public fun operationExport(vararg locale: Locale): String {
        return getMessageValue("operation.export", *locale)
    }

    public fun operationLimit(vararg locale: Locale): String {
        return getMessageValue("operation.limit", *locale)
    }

    public fun operationMultiple(vararg locale: Locale): String {
        return getMessageValue("operation.multiple", *locale)
    }

    public fun operationShowCount(vararg locale: Locale): String {
        return getMessageValue("operation.showCount", *locale)
    }

    public fun operationUpsert(vararg locale: Locale): String {
        return getMessageValue("operation.upsert", *locale)
    }

    public fun parameters(vararg locale: Locale): String {
        return getMessageValue("parameters", *locale)
    }

    public fun removeSubmit(vararg locale: Locale): String {
        return getMessageValue("remove.submit", *locale)
    }

    public fun resultCount(vararg locale: Locale): String {
        return getMessageValue("result.count", *locale)
    }

    public fun save(vararg locale: Locale): String {
        return getMessageValue("save", *locale)
    }

    public fun selectCollection(vararg locale: Locale): String {
        return getMessageValue("selectCollection", *locale)
    }

    public fun unknownQueryMethod(arg0: Any, vararg locale: Locale): String {
        return MessageFormat.format(getMessageValue("unknown.query.method", *locale), arg0)
    }

    public fun updateSubmit(vararg locale: Locale): String {
        return getMessageValue("update.submit", *locale)
    }

}
